src/diverse_seq/cluster.py

```diff
@@ -0,0 +1,618 @@
+                    distances[i, j] = distance
+                    distances[j, i] = distance
+
+                    if self._with_progress:
+                        self._progress.update(distance_task, advance=1)
+
+            return distances
+
+        # Compute distances in parallel
+        num_jobs = self._numprocs
+        futures = [
+            self._executor.submit(
+                compute_chunk_distances,
+                start,
+                num_jobs,
+                sketches,
+                self._k,
+                self._sketch_size,
+            )
+            for start in range(num_jobs)
+        ]
+
+        for future in futures:
+            start_idx, distances_chunk = future.result()
+            distances[start_idx::num_jobs] = distances_chunk[
+                start_idx::num_jobs
+            ].toarray()
+
+        # Make lower triangular matrix symmetric
+        distances = distances + distances.T - np.diag(distances.diagonal())
+
+        if self._with_progress:
+            self._progress.update(distance_task, completed=1, total=1)
+
+        return distances
+
+    def mash_sketches(self, records: list[DataMember]) -> list[BottomSketch]:
+        """Create sketch representations for a collection of sequence records.
+
+        Parameters
+        ----------
+        records : list[DataMember]
+            Sequence records.
+
+        Returns
+        -------
+        list[BottomSketch]
+            Sketches for each sequence.
+        """
+        if self._with_progress:
+            sketch_task = self._progress.add_task(
+                "[green]Generating Sketches",
+                total=len(records),
+            )
+
+        bottom_sketches = [None for _ in range(len(records))]
+
+        # Compute sketches in serial
+        if self._numprocs == 1:
+            for i, record in enumerate(records):
+                bottom_sketches[i] = mash_sketch(
+                    record,
+                    self._k,
+                    self._sketch_size,
+                    self._num_states,
+                    canonical=self._canonical,
+                )
+
+                if self._with_progress:
+                    self._progress.update(sketch_task, advance=1)
+
+            return bottom_sketches
+
+        # Compute sketches in parallel
+        futures_to_idx = {
+            self._executor.submit(
+                mash_sketch,
+                record,
+                self._k,
+                self._sketch_size,
+                self._num_states,
+                canonical=self._canonical,
+            ): i
+            for i, record in enumerate(records)
+        }
+
+        for future in as_completed(list(futures_to_idx)):
+            idx = futures_to_idx[future]
+            bottom_sketches[idx] = future.result()
+
+            if self._with_progress:
+                self._progress.update(sketch_task, advance=1)
+
+        return bottom_sketches
+
+
+def get_ordered_records(
```

Comment:
This is used to get the records for computing the pairwise mash distance matrix. For the agglomerative clustering step it assumes that the rows of the matrix are in the same order as `seq_names` (otherwise taxa are associated with incorrect sequences). This just returns an ordered list of records in the same order given by `seq_names`.